memgpt/client/streaming.py

```diff
@@ -0,0 +1,90 @@
+import json
+from typing import Generator
+
+import httpx
+from httpx_sse import SSEError, connect_sse
+
+from memgpt.constants import OPENAI_CONTEXT_WINDOW_ERROR_SUBSTRING
+from memgpt.errors import LLMError
+from memgpt.schemas.enums import MessageStreamStatus
+from memgpt.schemas.memgpt_message import (
+    FunctionCallMessage,
+    FunctionReturn,
+    InternalMonologue,
+)
+from memgpt.schemas.memgpt_response import MemGPTStreamingResponse
+
+
+def _sse_post(url: str, data: dict, headers: dict) -> Generator[MemGPTStreamingResponse, None, None]:
+
+    with httpx.Client() as client:
+        with connect_sse(client, method="POST", url=url, json=data, headers=headers) as event_source:
+
+            # Inspect for errors before iterating (see https://github.com/florimondmanca/httpx-sse/pull/12)
+            if not event_source.response.is_success:
+                # handle errors
+                from memgpt.utils import printd
+
+                printd("Caught error before iterating SSE request:", vars(event_source.response))
+                printd(event_source.response.read())
+
+                try:
+                    response_bytes = event_source.response.read()
+                    response_dict = json.loads(response_bytes.decode("utf-8"))
+                    error_message = response_dict["error"]["message"]
+                    # e.g.: This model's maximum context length is 8192 tokens. However, your messages resulted in 8198 tokens (7450 in the messages, 748 in the functions). Please reduce the length of the messages or functions.
+                    if OPENAI_CONTEXT_WINDOW_ERROR_SUBSTRING in error_message:
+                        raise LLMError(error_message)
+                except LLMError:
+                    raise
+                except:
+                    print(f"Failed to parse SSE message, throwing SSE HTTP error up the stack")
+                    event_source.response.raise_for_status()
+
+            try:
+                for sse in event_source.iter_sse():
+                    # if sse.data == OPENAI_SSE_DONE:
+                    # print("finished")
+                    # break
+                    if sse.data in [status.value for status in MessageStreamStatus]:
+                        # break
+                        # print("sse.data::", sse.data)
+                        yield MessageStreamStatus(sse.data)
```

Comment:
FYI @sarahwooders the reason we have duplicate `_sse_post` function here is so that we can yield a MemGPT specific type (`InternalMonologue`, `FunctionCallMessage`, etc)